File: app/api/filters.py
"""
Track filtering and smoothing endpoints.
"""
from fastapi import APIRouter
from app.core.database import get_conn, state, query
from app.core.track_filter import (
    compute_track_metrics, filter_tracks,
    apply_smoothing_to_db, smooth_trajectory,
    analyze_track_jitter, trim_track_jitter, auto_trim_all_jitter
)
import json

router = APIRouter(prefix="/filters", tags=["filters"])


@router.get("/metrics")
def get_track_metrics():
    """Get filtering metrics for all tracks (displacement, sinuosity, jitter, etc.)."""
    return compute_track_metrics(
        get_conn(), state['frame_width'], state['frame_height'])


@router.post("/apply")
def apply_filters(params: dict = None):
    """
    Compute which tracks to keep/hide based on filter parameters.
    Does NOT delete — returns filter results for the frontend.
    
    Body (all optional):
      min_displacement: float (default 30)
      min_frames: int (default 10)
      max_sinuosity: float (default 5.0)
      max_jitter: float (default 1.2)
      require_gate_crossing: bool (default false)
      remove_edge_only: bool (default true)
    """
    if params is None:
        params = {}

    results = filter_tracks(
        get_conn(), state['frame_width'], state['frame_height'],
        min_displacement=params.get('min_displacement', 30.0),
        min_frames=params.get('min_frames', 10),
        max_sinuosity=params.get('max_sinuosity', 5.0),
        max_jitter=params.get('max_jitter', 1.2),
        require_gate_crossing=params.get('require_gate_crossing', False),
        remove_edge_only=params.get('remove_edge_only', True),
    )

    keep = [tid for tid, r in results.items() if r['keep']]
    remove = {tid: r['reasons'] for tid, r in results.items() if not r['keep']}

    return {
        'total': len(results),
        'keep': len(keep),
        'remove': len(remove),
        'keep_ids': keep,
        'removed': remove,
    }


@router.post("/delete_tracks")
def delete_filtered_tracks(params: dict):
    """
    Permanently delete tracks by ID list.
    Body: { "track_ids": [1, 3, 6] }
    """
    track_ids = params.get('track_ids', [])
    if not track_ids:
        return {'deleted': 0}

    conn = get_conn()
    placeholders = ','.join('?' * len(track_ids))

    conn.execute(f"DELETE FROM observations WHERE track_id IN ({placeholders})", track_ids)
    conn.execute(f"DELETE FROM gate_crossings WHERE track_id IN ({placeholders})", track_ids)
    conn.execute(f"DELETE FROM tracklets WHERE global_track_id IN ({placeholders})", track_ids)
    conn.execute(f"DELETE FROM tracks WHERE track_id IN ({placeholders})", track_ids)
    conn.commit()

    # Update scene metadata
    new_count = conn.execute("SELECT COUNT(*) FROM tracks").fetchone()[0]
    conn.execute("UPDATE scene SET value=? WHERE key='total_vehicles'",
                 (str(new_count),))
    conn.commit()

    return {'deleted': len(track_ids), 'remaining': new_count}


@router.post("/smooth")
def smooth_trajectories(params: dict = None):
    """
    Apply moving-average smoothing to all trajectories.
    Body: { "window": 5 }  (default 5)
    """
    window = (params or {}).get('window', 5)
    updated = apply_smoothing_to_db(get_conn(), window=window)
    return {'smoothed': updated, 'window': window}


# ──────────────────────────────────────────────────────────
# TRAJECTORY TRIMMING
# ──────────────────────────────────────────────────────────

@router.get("/jitter_analysis/{track_id}")
def get_jitter_analysis(track_id: int, threshold: float = 1.0):
    """
    Analyze a single track's trajectory for jittery segments.
    Returns segment breakdown (clean vs jittery regions).
    """
    result = analyze_track_jitter(get_conn(), track_id,
                                  jitter_threshold=threshold)
    if result is None:
        return {'error': 'Track not found'}
    return result


@router.post("/trim_track")
def trim_track(params: dict):
    """
    Trim a track's trajectory to keep only a specific range.
    Body: { "track_id": 42, "keep_start": 5, "keep_end": 180 }
    """
    track_id = params.get('track_id')
    keep_start = params.get('keep_start', 0)
    keep_end = params.get('keep_end')
    if track_id is None or keep_end is None:
        return {'error': 'track_id and keep_end are required'}
    return trim_track_jitter(get_conn(), track_id, keep_start, keep_end)


@router.post("/auto_trim_jitter")
def auto_trim_jitter(params: dict = None):
    """
    Automatically trim jittery ends from ALL trajectories.
    Keeps the longest clean segment of each track.
    Body: { "jitter_threshold": 1.0, "min_clean_length": 5 }
    """
    params = params or {}
    threshold = params.get('jitter_threshold', 1.0)
    min_clean = params.get('min_clean_length', 5)
    return auto_trim_all_jitter(get_conn(), jitter_threshold=threshold,
                                min_clean_length=min_clean)


@router.post("/preview_trim_jitter")
def preview_trim_jitter(params: dict = None):
    """
    Preview what auto-trim would do without applying changes.
    Returns per-track trim analysis.
    """
    params = params or {}
    threshold = params.get('jitter_threshold', 1.0)
    min_clean = params.get('min_clean_length', 5)

    conn = get_conn()
    tracks = conn.execute("SELECT track_id, trajectory_json FROM tracks").fetchall()

    preview = []
    total_would_trim = 0
    total_points_removed = 0

    for t in tracks:
        tid = t[0]
        traj = json.loads(t[1]) if t[1] else []
        if len(traj) < 6:
            continue

        from app.core.track_filter import find_jittery_segments
        segments = find_jittery_segments(traj, jitter_threshold=threshold,
                                         min_clean_length=min_clean)

        # Find longest clean segment
        best_seg = None
        best_len = 0
        for seg in segments:
            if seg['type'] == 'clean':
                seg_len = seg['end'] - seg['start'] + 1
                if seg_len > best_len:
                    best_len = seg_len
                    best_seg = seg

        if best_seg and not (best_seg['start'] == 0 and best_seg['end'] == len(traj) - 1):
            points_removed = len(traj) - best_len
            total_would_trim += 1
            total_points_removed += points_removed
            preview.append({
                'track_id': tid,
                'total_points': len(traj),
                'keep_start': best_seg['start'],
                'keep_end': best_seg['end'],
                'keep_points': best_len,
                'removing_points': points_removed,
                'segments': segments,
            })

    return {
        'total_tracks': len(tracks),
        'would_trim': total_would_trim,
        'unchanged': len(tracks) - total_would_trim,
        'total_points_removed': total_points_removed,
        'details': preview[:200],  # Cap for response size
    }

File: static/js/filters.js
// ── Track Filtering & Smoothing ──
var pendingDeleteIds = [];
var hiddenTrackIds = new Set();  // tracks visually hidden by filter preview

async function smoothTrajectories(window) {
  if (!confirm('Smooth all trajectories with window=' + window + '?\nThis modifies the .traf file.')) return;
  
  var res = await fetch('/api/filters/smooth', {
    method: 'POST',
    headers: {'Content-Type': 'application/json'},
    body: JSON.stringify({window: window})
  });
  var data = await res.json();
  alert('Smoothed ' + data.smoothed + ' trajectories (window=' + data.window + ')');
  
  // Reload trajectories
  allTrajectories = await fetchJSON('/api/trajectories?stationary=0');
  render();
}

async function previewFilter() {
  var params = {
    min_displacement: parseFloat(document.getElementById('fMinDisp').value) || 30,
    min_frames: parseInt(document.getElementById('fMinFrames').value) || 10,
    max_sinuosity: parseFloat(document.getElementById('fMaxSin').value) || 5.0,
    max_jitter: parseFloat(document.getElementById('fMaxJitter').value) || 1.2,
    remove_edge_only: document.getElementById('fEdgeOnly').checked,
    require_gate_crossing: document.getElementById('fGateRequired').checked,
  };

  var res = await fetch('/api/filters/apply', {
    method: 'POST',
    headers: {'Content-Type': 'application/json'},
    body: JSON.stringify(params)
  });
  var data = await res.json();

  // Show preview
  var el = document.getElementById('filterPreview');
  var removedList = Object.entries(data.removed);

  if (removedList.length === 0) {
    el.innerHTML = '<div style="color:var(--green);padding:6px;">✓ All ' + data.total + ' tracks pass filters. Nothing to remove.</div>';
    document.getElementById('btnDeleteFiltered').style.display = 'none';
    hiddenTrackIds.clear();
    render();
    return;
  }

  var html = '<div style="color:var(--yellow);padding:4px 0;">Would remove ' + data.remove + ' of ' + data.total + ' tracks:</div>';
  html += '<div style="max-height:150px;overflow-y:auto;margin:4px 0;">';
  removedList.forEach(function(item) {
    var tid = item[0], reasons = item[1];
    html += '<div style="display:flex;justify-content:space-between;padding:2px 0;border-bottom:1px solid var(--border);">';
    html += '<span style="color:var(--accent);">#' + tid + '</span>';
    html += '<span style="color:var(--text-muted);font-size:10px;">' + reasons.join(', ') + '</span>';
    html += '</div>';
  });
  html += '</div>';
  el.innerHTML = html;

  // Store IDs for deletion and visually hide them
  pendingDeleteIds = removedList.map(function(item) { return parseInt(item[0]); });
  hiddenTrackIds = new Set(pendingDeleteIds);

  document.getElementById('btnDeleteFiltered').style.display = 'block';
  document.getElementById('btnDeleteFiltered').textContent = '🗑 Delete ' + pendingDeleteIds.length + ' Tracks';

  // Re-render with hidden tracks
  render();
}

async function applyFilterDelete() {
  if (!pendingDeleteIds.length) return;
  if (!confirm('Permanently delete ' + pendingDeleteIds.length + ' tracks?\nThis cannot be undone.')) return;

  var res = await fetch('/api/filters/delete_tracks', {
    method: 'POST',
    headers: {'Content-Type': 'application/json'},
    body: JSON.stringify({track_ids: pendingDeleteIds})
  });
  var data = await res.json();

  alert('Deleted ' + data.deleted + ' tracks. ' + data.remaining + ' remaining.');

  // Reload everything
  pendingDeleteIds = [];
  hiddenTrackIds.clear();
  document.getElementById('btnDeleteFiltered').style.display = 'none';
  document.getElementById('filterPreview').innerHTML = '';

  allTracks = await fetchJSON('/api/tracks');
  allTrajectories = await fetchJSON('/api/trajectories?stationary=0');
  summary = await fetchJSON('/api/summary');

  // Recompute gate crossings
  try { window.gates = await fetchJSON('/api/gates/count_summary'); } catch(e) {}

  renderDashboard();
  renderTrackList(allTracks);
  renderGateList();
  render();
}

async function deleteSelectedTrack() {
  if (!selectedTrack) {
    alert('Select a track first in the TRACKS tab.');
    return;
  }
  if (!confirm('Delete track #' + selectedTrack + '?')) return;

  await fetch('/api/filters/delete_tracks', {
    method: 'POST',
    headers: {'Content-Type': 'application/json'},
    body: JSON.stringify({track_ids: [selectedTrack]})
  });

  selectedTrack = null;
  allTracks = await fetchJSON('/api/tracks');
  allTrajectories = await fetchJSON('/api/trajectories?stationary=0');
  summary = await fetchJSON('/api/summary');
  renderDashboard();
  renderTrackList(allTracks);
  render();
}


// ──────────────────────────────────────────────────────────
// TRAJECTORY TRIMMING
// ──────────────────────────────────────────────────────────

var pendingTrimData = null;
var selectedTrackTrimInfo = null;

async function previewTrimJitter() {
  var threshold = parseFloat(document.getElementById('fTrimThreshold').value) || 1.0;
  var minClean = parseInt(document.getElementById('fTrimMinClean').value) || 5;

  document.getElementById('trimPreview').innerHTML = '<div style="color:var(--text-muted);">Analyzing...</div>';

  var res = await fetch('/api/filters/preview_trim_jitter', {
    method: 'POST',
    headers: {'Content-Type': 'application/json'},
    body: JSON.stringify({jitter_threshold: threshold, min_clean_length: minClean})
  });
  var data = await res.json();
  pendingTrimData = data;

  var el = document.getElementById('trimPreview');
  if (data.would_trim === 0) {
    el.innerHTML = '<div style="color:var(--green);padding:6px;">✓ No jittery segments found. All trajectories are clean.</div>';
    document.getElementById('btnApplyTrim').style.display = 'none';
    return;
  }

  var html = '<div style="color:var(--yellow);padding:4px 0;">Would trim ' + data.would_trim + ' of ' + data.total_tracks + ' tracks</div>';
  html += '<div style="color:var(--text-muted);padding:2px 0;">Total points to remove: ' + data.total_points_removed + '</div>';
  html += '<div style="max-height:150px;overflow-y:auto;margin:4px 0;">';
  data.details.forEach(function(d) {
    html += '<div style="display:flex;justify-content:space-between;padding:2px 0;border-bottom:1px solid var(--border);">';
    html += '<span style="color:var(--accent);">#' + d.track_id + '</span>';
    html += '<span style="color:var(--text-muted);font-size:10px;">' + d.total_points + '→' + d.keep_points + ' pts (−' + d.removing_points + ')</span>';
    html += '</div>';
  });
  html += '</div>';
  el.innerHTML = html;

  document.getElementById('btnApplyTrim').style.display = 'block';
  document.getElementById('btnApplyTrim').textContent = '✂ Trim ' + data.would_trim + ' Tracks';
}

async function applyTrimJitter() {
  if (!pendingTrimData || pendingTrimData.would_trim === 0) return;
  if (!confirm('Trim jittery ends from ' + pendingTrimData.would_trim + ' tracks?\nThis modifies the .traf file.')) return;

  var threshold = parseFloat(document.getElementById('fTrimThreshold').value) || 1.0;
  var minClean = parseInt(document.getElementById('fTrimMinClean').value) || 5;

  var res = await fetch('/api/filters/auto_trim_jitter', {
    method: 'POST',
    headers: {'Content-Type': 'application/json'},
    body: JSON.stringify({jitter_threshold: threshold, min_clean_length: minClean})
  });
  var data = await res.json();

  alert('Trimmed ' + data.tracks_trimmed + ' tracks. Removed ' + data.total_points_removed + ' jittery points.');

  // Reset and reload
  pendingTrimData = null;
  document.getElementById('btnApplyTrim').style.display = 'none';
  document.getElementById('trimPreview').innerHTML = '';

  allTracks = await fetchJSON('/api/tracks');
  allTrajectories = await fetchJSON('/api/trajectories?stationary=0');
  summary = await fetchJSON('/api/summary');

  try { window.gates = await fetchJSON('/api/gates/count_summary'); } catch(e) {}

  renderDashboard();
  renderTrackList(allTracks);
  renderGateList();
  render();
}

// Analyze selected track for jitter when it changes
async function analyzeSelectedTrackJitter() {
  var trimSection = document.getElementById('selectedTrackTrim');
  var infoEl = document.getElementById('trackTrimInfo');
  var btnTrim = document.getElementById('btnTrimSelected');

  if (!selectedTrack) {
    trimSection.style.display = 'none';
    selectedTrackTrimInfo = null;
    return;
  }

  trimSection.style.display = 'block';
  infoEl.innerHTML = '<span style="color:var(--text-muted);">Analyzing #' + selectedTrack + '...</span>';

  var threshold = parseFloat(document.getElementById('fTrimThreshold').value) || 1.0;

  try {
    var data = await fetchJSON('/api/filters/jitter_analysis/' + selectedTrack + '?threshold=' + threshold);
    selectedTrackTrimInfo = data;

    if (data.error) {
      infoEl.innerHTML = '<span style="color:var(--red);">' + data.error + '</span>';
      btnTrim.style.display = 'none';
      return;
    }

    // Show segment breakdown
    var html = '<div style="margin-bottom:4px;">Track #' + data.track_id + ' — ' + data.total_points + ' points</div>';
    var hasJitter = false;

    data.segments.forEach(function(seg) {
      var len = seg.end - seg.start + 1;
      var color = seg.type === 'clean' ? 'var(--green)' : 'var(--red)';
      var icon = seg.type === 'clean' ? '━' : '〰';
      html += '<div style="display:flex;gap:6px;padding:1px 0;">';
      html += '<span style="color:' + color + ';">' + icon + '</span>';
      html += '<span>' + seg.type + ': pts ' + seg.start + '–' + seg.end + ' (' + len + ')</span>';
      html += '</div>';
      if (seg.type === 'jittery') hasJitter = true;
    });

    infoEl.innerHTML = html;
    btnTrim.style.display = hasJitter ? 'block' : 'none';

  } catch(e) {
    infoEl.innerHTML = '<span style="color:var(--red);">Analysis failed</span>';
    btnTrim.style.display = 'none';
  }
}

async function trimSelectedTrack() {
  if (!selectedTrackTrimInfo || !selectedTrack) return;

  // Find longest clean segment
  var best = null, bestLen = 0;
  selectedTrackTrimInfo.segments.forEach(function(seg) {
    if (seg.type === 'clean') {
      var len = seg.end - seg.start + 1;
      if (len > bestLen) { bestLen = len; best = seg; }
    }
  });

  if (!best) {
    alert('No clean segment found — consider deleting this track instead.');
    return;
  }

  var removing = selectedTrackTrimInfo.total_points - bestLen;
  if (!confirm('Trim track #' + selectedTrack + '?\nKeeping points ' + best.start + '–' + best.end + ' (' + bestLen + ' pts)\nRemoving ' + removing + ' jittery points.')) return;

  var res = await fetch('/api/filters/trim_track', {
    method: 'POST',
    headers: {'Content-Type': 'application/json'},
    body: JSON.stringify({track_id: selectedTrack, keep_start: best.start, keep_end: best.end})
  });
  var data = await res.json();

  if (data.error) {
    alert('Error: ' + data.error);
    return;
  }

  alert('Trimmed: ' + data.old_points + ' → ' + data.new_points + ' points.');

  // Reload
  allTracks = await fetchJSON('/api/tracks');
  allTrajectories = await fetchJSON('/api/trajectories?stationary=0');
  render();
  analyzeSelectedTrackJitter();
}
